demos/ret2spec_cyclic.cc

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright 2019 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Demonstration of ret2spec that exploits the fact that return stack buffers
+ * have limited size and they can be rewritten by recursive invocations of
+ * another function.
+ * We have two functions that we named after their constant return values. First
+ * the returns_true function invokes kRecursionDepth times itself and in the
+ * deepest invocation it calls returns_false function. Returns_false function
+ * invokes kRecursionDepth times itself. All returns of the returns_false
+ * function are predicted correctly, but returns of returns_true function are
+ * mispredicted to the return address of the returns_false function, because all
+ * RSB pointers were rewritten by returns_false invocations. We steer those
+ * mispredictions to an unreachable code path with microarchitectural side
+ * effects.
+ **/
+
+#include <array>
+#include <cstring>
+#include <iostream>
+#include <vector>
+
+#include "cache_sidechannel.h"
+#include "instr.h"
+
+const char *private_data = "It's a s3kr3t!!!";
+
+// Recursion depth should be equal or greater than the RSB size, but not
+// excessively high because of the possibility of stack overflow.
+constexpr size_t kRecursionDepth = 64;
+constexpr size_t kCacheLineSize = 16;
```

Comment:
This one's surprising. Is this size in bytes? It's usually 64.

Reply:
Fixed.